services/database/DAO.js

```diff
@@ -0,0 +1,349 @@
+/**
+ * This represents the high level (psudeo-) DAO used for interacting with postgres.
+ */
+
+const { Client } = require('pg');
+
+class AbstractDAO {
+    /**
+     * This abstract class is provided primarily for testing
+     * The PGDAO overrides with an actual Postgres implementation of the DAO while
+     * the InMemDAO is exclusively in memory. The AbstractDAO provides a common interface
+     */
+    noAbstract() {
+        throw new Error("Cannot run request on the AbstractDAO. Use one of its children.");
+    }
+
+    async insert(tableSchema, row) {
+        /**
+         * Inserts a row into a table.
+         * This requires that
+         * 1. the full key is contained within the row
+         * 2. all non-nullable fields are within the row
+         * 3. the row does not already exist within the table
+         */
+        this.noAbstract();
```

Comment:
Seems unnecessary to write the this.noAbstract() since you can't instantiate the class anyways lol